File: python/app/fraud.py
from typing import Dict, Tuple


def evaluate_fraud_signals(payload: Dict[str, object]) -> Tuple[int, str, Dict[str, object]]:
    amount = int(payload["amount"])
    currency = payload.get("currency", "USD")
    signals: Dict[str, object] = {}

    score = 10
    if amount > 50000:
        score += 40
        signals["high_amount"] = True

    if currency not in {"USD", "EUR", "GBP"}:
        score += 20
        signals["unsupported_currency"] = currency

    if payload.get("customer_id") is None:
        score += 10
        signals["guest_checkout"] = True

    decision = "allow"
    if score >= 60:
        decision = "block"
    elif score >= 40:
        decision = "review"

    return score, decision, signals

File: typescript/src/common/fraud.ts
import { PaymentIntent } from "./store.js";

export type FraudDecision = "allow" | "review" | "block";

export const evaluateFraud = (intent: PaymentIntent) => {
  let score = 10;
  const signals: Record<string, unknown> = {};

  if (intent.amount > 50000) {
    score += 40;
    signals.high_amount = true;
  }
  if (!["USD", "EUR", "GBP"].includes(intent.currency)) {
    score += 20;
    signals.unsupported_currency = intent.currency;
  }
  if (!intent.customerId) {
    score += 10;
    signals.guest_checkout = true;
  }

  let decision: FraudDecision = "allow";
  if (score >= 60) decision = "block";
  else if (score >= 40) decision = "review";

  return { score, decision, signals };
};
